### PR TITLE
DAOS-13144 common: fix to free lmdb memory

### PR DESCRIPTION
**Pull request by wangshilong**
1. In vos_self_fini_locked() bio_nvme_fini() need be called after sys db cleanup, because we need bio_nvme_configure() to check if MD on SSD enabled or not.
2. in lmm_db_fini() env need be destroyed after dbi closed to avoid use-after-free.
3. refactored lmm_db_init_ex() a bit to avoid potential memory leaks if lmm_db_open_create() return NOEXIST but it is not file not existed case.
4. add missed bio_nvme_fini() in the engine mode.

### CODE REVIEW

File: src/vos/vos_common.c [hunk 1]
```diff
@@ -804,11 +804,11 @@ vos_self_fini_locked(void)
 		self_mode.self_xs_ctxt = NULL;
 	}
 
-	vos_self_nvme_fini();
 	if (!bio_nvme_configured(SMD_DEV_TYPE_META))
 		vos_db_fini();
 	else
 		lmm_db_fini();
+	vos_self_nvme_fini();
```
Comment (NiuYawei) [inline src/vos/vos_common.c:811]: bio_nvme_fini() is missed in engine mode, could you add it back in this PR? Thanks.

Comment (wangshilong) [reply src/vos/vos_common.c:811]: sure

**Review by NiuYawei — DISMISSED** [review]

File: src/common/sys_lmdb.c [hunk 2]
```diff
@@ -430,11 +427,13 @@ lmm_db_fini(void)
...
-			mdb_env_close(lmm_db.db_env);
 		if (lmm_db.db_destroy_db)
 			lmm_db_unlink(&lmm_db.db_pub);
-		mdb_dbi_close(lmm_db.db_env, lmm_db.db_dbi);
+		if (lmm_db.db_env) {
+			if (lmm_db.db_dbi_valid)
+				mdb_dbi_close(lmm_db.db_env, lmm_db.db_dbi);
+			mdb_env_close(lmm_db.db_env);
+		}
 		free(lmm_db.db_file);
```
Comment (jolivier23) [inline src/common/sys_lmdb.c:437]: we should use D_ASPRINTF and D_FREE for these

**Review by jolivier23 — DISMISSED** [review]

File: src/common/sys_lmdb.c (hunk 2 above)
Comment (wangshilong) [reply src/common/sys_lmdb.c]: Sure

**Review by jolivier23 — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

File: src/common/sys_lmdb.c [hunk 3]
```diff
@@ -457,17 +455,18 @@ lmm_db_init_ex(const char *db_path, const char *db_name, bool force_create, bool
 	if (rc != ABT_SUCCESS)
 		return -DER_NOMEM;
 
-	rc = asprintf(&lmm_db.db_path, "%s", db_path);
-	if (rc < 0) {
+	D_ASPRINTF(lmm_db.db_path, "%s", db_path);
+	if (lmm_db.db_path == NULL) {
 		D_ERROR("Generate sysdb path failed. %d\n", rc);
-		return -DER_NOMEM;
+		rc = -DER_NOMEM;
```
Comment (jolivier23) [inline src/common/sys_lmdb.c:461]: This error message is unecessary.  D_ASPRINTF will print an error

File: src/common/sys_lmdb.c [hunk 4]
```diff
@@ -457,17 +455,18 @@ lmm_db_init_ex(const char *db_path, const char *db_name, bool force_create, bool
...
 	}
 
 	if (!db_name)
 		db_name = SYS_DB_NAME;
 
-	rc = asprintf(&lmm_db.db_file, "%s/%s", lmm_db.db_path, db_name);
-	if (rc < 0) {
+	D_ASPRINTF(lmm_db.db_file, "%s/%s", lmm_db.db_path, db_name);
+	if (lmm_db.db_file == NULL) {
 		D_ERROR("Generate sysdb filename failed. %d\n", rc);
```
Comment (jolivier23) [inline src/common/sys_lmdb.c:470]: this message is no longer needed.

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**